File: prepare_data/prepare_data.jl
include("../BBDD/mongo_connection.jl")
using DataFrames
using CSV
using Dates
import Tables: rows


function log(x) println("$(now()) | $x") end

csv_filename = "tp_data_final.csv"

db_testing = "testing"

col_imdb_complete = "imdb_complete"
col_tmdb_complete = "tmdb_complete"
col_comments_rating = "collaborative_db"
col_titlebasics = "imdbTitleBasics"
col_titleratings = "imdbTitleRatings"
col_namebasics = "imdbNameBasics"
col_datatp = "data_tp_final"


imdbs_list = [
    "tt10133702",
    "tt10333266",
    "tt10350420",
    "tt10399902",
    "tt10402396",
    "tt10469410",
    "tt10712472",
    "tt11244166",
    "tt11257606",
    "tt11281192",
]
    
mongo = Mongo("localhost", 27017)


function complete_imdb_nan(df)
    for col_name in names(df)
        try
            df[:, col_name] = replace.(df[:, col_name], "\\N" => missing)
            df[:, col_name] = replace.(df_items[:, col_name], "\\\\N" => missing)
        catch e
            nothing
        end
    end
    return df
end



function isnull(value)
    if ismissing(value) || value===nothing || value=="" || value=="nothing"
        return true
    end
    return false
end


function set_company_names(value)
    if value === nothing
        return nothing
    end
    companies = String[]
    for i in value
        push!(companies, String(i["Name"]))
    end
    return companies
end


function set_country(value)
    if value === nothing
        return nothing
    end
    return value[1]
end


function set_string_array(value)
    if value === nothing
        return nothing
    end
    values_list = String[]
    for i in value
        push!(values_list, String(i))
    end
    return values_list
end


function get_data_from_mongo(database::String, collection::String, query::Dict, projection::Dict)

    # log("Searching data in '$database.$collection'.")
    
    pipeline = [
        Dict( "\$match" => query ),
        Dict( "\$count" => "count" )
    ]
    
    result_count = run_aggregate(mongo, database, collection, pipeline)
    if length(result_count) == 0
        log("\tNo items found in '$database.$collection'.\n")
        return DataFrame
    end
    total_items = result_count[1]["count"]

    step = 1000000
    # log("\tThere are $total_items items to get.")

    n_got = 0
    df_items = DataFrame()

    for start in 0 : step : total_items+1
        pipeline = [
            Dict( "\$match" => query ),
            Dict( "\$skip" => start ),
            Dict( "\$limit" => step ),
            Dict( "\$project" => projection ),
        ]
        
        items = run_aggregate(mongo, database, collection, pipeline)
        
        df_current = DataFrame(items)
        # df_current = vcat(DataFrame.(items)...)
        
        df_items = vcat( df_items, df_current )
        
        n_got += length(items)
        # log("\t\t$n_got / $total_items")
    end
    
    if "imdbId" in names(df_items)
        dropmissing!(df_items, :imdbId);
    end

    df_items = complete_imdb_nan(df_items)

    # log("\t$(nrow(df_items)) total items got.\n")
    return df_items
end


# #### Imdb TitleBasics data
function get_title_basics()
    #=
        Returns DataFrame with columns: imdbId, type, year, duration, genres
    =#
    # query = Dict()
    query = Dict( "imdbId" => Dict( "\$in" => imdbs_list ) )
    projection = Dict(
        "_id" => 0,
        "imdbId" => "\$imdbId",
        "type" => "\$titleType",
        "year" => "\$startYear",
        "duration" => "\$runtimeMinutes",
        "genres" => "\$genres",
    )
    df_titlebasics = get_data_from_mongo(db_testing, col_titlebasics, query, projection)
    transform!(df_titlebasics, :genres .=> ByRow(x -> [ String(i) for i in split(x, ",") ]) .=> :genres)
    log("$(nrow(df_titlebasics)) items from $col_titlebasics ") # \n$(first(df_titlebasics, 2))\n")

    # log("\n\n$df_titlebasics\n\n")
    return df_titlebasics
end


# #### Imdb TitleRatings data
function get_title_ratings(df::DataFrame)
    #=
        Arguments
            df (DataFrame) with columns (at least) imdbId
        Returns with original dataframe with added columns rating, votes
    =#
    # query = Dict()
    query = Dict( "imdbId" => Dict( "\$in" => imdbs_list ) )
    projection = Dict(
        "_id" => 0,
        "imdbId" => "\$imdbId",
        "rating" => "\$averageRating",
        "votes" => "\$numVotes",
    )
    df_titleratings = get_data_from_mongo(db_testing, col_titleratings, query, projection)
    log("$(nrow(df_titleratings)) items from $col_titleratings ") # \n$(first(df_titleratings, 2))\n")
    df = leftjoin(df, df_titleratings, on=:imdbId) # Add columns rating, votes
    # log("\t -->> $(nrow(df)) ITEMS <<--\n")
    # log("\n\n$df\n\n")
    return df
end


# #### Letterboxd reviews data 
# Buscar ultima fecha (createdAt) y filtrar asi
function get_reviews(df::DataFrame)
    #=
        Arguments
            df (DataFrame) with columns (at least) imdbId
        Returns with original dataframe with added column reviews
    =#
    # query = Dict()
    query = Dict( "imdb_id" => Dict( "\$in" => imdbs_list ) )
    projection = Dict(
        "_id" => 0,
        "imdbId" => "\$imdb_id",
        "reviews" => "\$reviews_analysis.compound",
    )
    df_collaborativedb = get_data_from_mongo(db_testing, col_comments_rating, query, projection)
    df_collaborativedb = combine(groupby(df_collaborativedb, :imdbId), :reviews => maximum => :reviews) # EXPLICAR
    log("$(nrow(df_collaborativedb)) items from $col_comments_rating ") # \n$(first(df_collaborativedb, 2))\n")
    df = leftjoin(df, df_collaborativedb, on=:imdbId)
    # log("\t -->> $(nrow(df)) ITEMS <<--\n")
    # log("\n\n$df\n\n")
    return df
end


# #### Imdb NameBasics data (directors)
function get_name_basics(df::DataFrame)
    #=
        Arguments
            df (DataFrame) with columns (at least) imdbId
        Returns with original dataframe with added column directors
    =#
    query = Dict(
        "\$or" => [ Dict( "knownForTitles" => Dict( "\$regex" => value ) ) for value in imdbs_list ]
    )
    projection = Dict(
        "_id" => 0,
        "profession" => "\$primaryProfession",
        "name" => "\$primaryName",
        "titles" => "\$knownForTitles"
    )
    
    # df_namebasics = get_data_from_mongo(db_testing, col_namebasics, query, projection)
    # CSV.write("names_test.csv", df_namebasics)
    df_namebasics = DataFrame( CSV.read("names_test.csv", DataFrame) )
    
    df_namebasics = filter(row -> occursin( "director" , row.profession ) , df_namebasics)
    df_namebasics = df_namebasics[ : , [:name, :titles] ]
    transform!(df_namebasics, :titles .=> ByRow(x -> [ String(i) for i in split(x, ",") ]) .=> :titles)
    df_namebasics = DataFrame( [ ( name=row["name"], imdbId=val ) for row in eachrow(df_namebasics) for val in row[:titles] ] )
    df_namebasics = filter(row -> row.imdbId in imdbs_list , df_namebasics)
    df_directors = DataFrame()
    for i in groupby( df_namebasics, :imdbId )
        names_vector = [ String(_name) for _name in i[:, :name] ]
        imdb = i[1,"imdbId"]
        push!(df_directors, ( directors=names_vector, imdbId=imdb ))
    end
    df = leftjoin(df, df_directors, on=:imdbId)

    # log("\t -->> $(nrow(df)) ITEMS <<--\n")
    # log("\n\n$df\n\n")
    return df
end


# Imdb complete data 
function get_imdb_complete(df::DataFrame)
    #=
        Arguments
            df (DataFrame) with columns (at least) imdbId
        Returns with original dataframe with added columns cast, country, releaseDate, keywords, companies
    =#
    # query = Dict()
    query = Dict( "Id" => Dict( "\$in" => imdbs_list ) )
    projection = Dict(
        "_id" => 0,
        "imdbId" => "\$Id",
        "cast" => "\$Cast",
        "country" => "\$Country",
        "releaseDate" => "\$ReleaseDate",
        "keywords" => "\$Keywords",
        "companies" => "\$Companies",
    )
    df_imdb_complete = get_data_from_mongo(db_testing, col_imdb_complete, query, projection)

    df_imdb_complete = transform(df_imdb_complete, :companies .=> ByRow(x -> set_company_names(x)) .=> :companies)
    df_imdb_complete = transform(df_imdb_complete, :country .=> ByRow(x -> set_country(x)) .=> :country)

    for col_name in [ "cast", "keywords" ]
        df_imdb_complete = transform(df_imdb_complete, col_name .=> ByRow(x -> set_string_array(x)) .=> col_name)
    end

    log("$(nrow(df_imdb_complete)) items from $col_imdb_complete ") # \n$(first(df_imdb_complete, 2))\n")

    df = leftjoin(df, df_imdb_complete, on=:imdbId) # Add columns: cast, country, releaseDate, keywords, companies

    # log("\t -->> $(nrow(df)) ITEMS <<--\n")
    # log("\n\n$df\n\n")
    return df
end


# Tmdb complete data
function get_tmdb_complete()
    # query = Dict( "imdbId" => Dict( "\$ne" => nothing ) )
    query = Dict( "imdbId" => Dict( "\$in" => imdbs_list ) )
    projection = Dict(
        "_id" => 0,
        "imdbId" => 1,
        "releaseDate" => "\$ReleaseDate",
        "votes" => "\$vote_count",
        "rating" => "\$vote_average",
        "duration" => "\$Duration",
        "genres" => "\$Genres",
        "country" => "\$Country",
        "keywords" => "\$Keywords",
        "cast" => "\$Cast",
        "companies" => "\$Companies",
        "directors" => "\$Directors"
    )
    df_tmdb = get_data_from_mongo(db_testing, col_tmdb_complete, query, projection)

    df_tmdb = unique(df_tmdb, :imdbId)

    df_tmdb = transform(df_tmdb, :companies .=> ByRow(x -> set_company_names(x)) .=> :companies)
    df_tmdb = transform(df_tmdb, :country .=> ByRow(x -> set_country(x)) .=> :country)
    for col_name in [ "genres", "companies", "cast", "directors", "keywords" ]
        df_tmdb = transform(df_tmdb, col_name .=> ByRow(x -> set_string_array(x)) .=> col_name)
    end

    log("$(nrow(df_tmdb)) items from $col_tmdb_complete ") # \n$(first(df_tmdb, 2))\n")

    return df_tmdb
end


function complete_data(df, df_tmdb)

    cols = names(df)

    log("Completing data for $(cols) rows:\n")

    fields = [
		"releaseDate",
		"votes",
		"rating",
		"duration",
        "genres",
        "country",
        "keywords",
        "cast",
        "companies",
        "directors",
    ]

    index = 0
    for field in fields
        index += 1
        log("\tField '$field' $(index)/$(length(fields))")
        # log("$(names(df))")

        # log("\t\tGetting null data from df")
        null_data = df[ [  isnull(i) for i in df[:, field] ] , : ]
        if isempty(null_data)
            log("\t\t\tNo null data")
            continue
        end

        # log("\t\tGetting not null data from df")
        not_null_data = df[ [  !(isnull(i)) for i in df[:,field] ] , : ]

        df_tmdb_info = df_tmdb[ : , ["imdbId", field] ]
        # log("\t\tGetting not null data from tmdb df")
        df_tmdb_info = df_tmdb_info[ [  !(isnull(i)) for i in df_tmdb_info[:,field] ] , : ]

        # log("\t\tRemoving field $field from df")
        select!(null_data, Not([field]))
        
        # log("\t\tJoin data")
        null_data = leftjoin(null_data, df_tmdb_info, on=:imdbId)

        # log("\t\tConcat data")
        df = vcat( null_data[:, cols] , not_null_data[:, cols] )

        # println("\n\n")
    end
	
    return df
end


function insert_to_mongo(df)

    log("Inserting data: $(nrow(df)) rows:\n")
    items = rows(df) |> collect

    items_dict = []
    for item in items
        a = Dict(names(item) .=> values(item))
        # print(a)
        push!(items_dict, a)
    end

    log("\tInserting to mongo...")

    # insert_many_contents(mongo, db_testing, col_datatp, items_dict)
    for item in items_dict
        try
            insert_content(mongo, db_testing, col_datatp, item)
        catch e
            println("\n\nERROR: $e \n$item\n\n")
            break
        end
    end
end


function set_nothing_to_missing(df::DataFrame)

    println("\n\n")
    log("Setting nothing values to missing.")
    columns = names(df)[1:7]
    # columns = [ "duration" ]

    # df = df[:, [ "duration", "year", "type", "imdbId", "reviews", "rating" ] ]
    println("\n\n$( df[ : , columns ] )\n\n")

    index, n = 1, length(columns)
    for column in columns
        log("\tColumn '$column' $index/$n")
        index += 1
        df_nulls = nothing
        try
            df_nulls = df[ [ isnull(i) for i in df[:,column] ] , : ]
        catch
            print("\n\nERROR GETTING NULL VALUES FOR COLUMN:\n$(df[:,column])\n\n")
            continue
        end
        # println("\nNULL VALUES\n$df_nulls\n")
        if isempty(df_nulls)
            log("\t\tNo nulls values\n\n")
            continue
        end
        println("\t\tNULL VALUES\n$df_nulls\n\n")

        # df_not_nulls = df[ [  i!="nothing" for i in df[:,column] ] , : ]
        # println("\nNOT NULL VALUES\n$df_not_nulls\n")

        # df_nulls[ ! , column ] .= missing
        # println("\nNEW NULL VALUES\n$df_nulls\n")

        # df = vcat( df_not_nulls , df_nulls )
    end

    # println("\n\n$df\n\n")

	return df
end


function main()

    # df = DataFrame()
|
    df = get_title_basics()
    df = get_title_ratings(df)
    df = get_reviews(df)
    df = get_name_basics(df)
    df = get_imdb_complete(df)
    df_tmdb = get_tmdb_complete()
    
    # CSV.write("out.csv", df, transform=(col, val) -> something(val, missing))
    # CSV.write("out2.csv", df_tmdb, transform=(col, val) -> something(val, missing))
    
    df = complete_data(df, df_tmdb)

    # df = df.fillna(value=nan).replace(to_replace=nan, value=None)
    # df = df.replace(to_replace="", value=nothing)
    # df = df.replace(to_replace="\\N", value=nothing)
    # df = df.replace(to_replace="\\\\N", value=nothing)

    # CSV.write("out3.csv", df, transform=(col, val) -> something(val, missing))
    
    df = set_nothing_to_missing(df)

    # insert_to_mongo(df)

    # df = df[ [  !(isnull(i)) for i in df[:, :rating] ] , : ]
    # CSV.write(csv_filename, df)

    # println("\n\n$df\n\n")

end


if abspath(PROGRAM_FILE) == @__FILE__
    main()
end


#=

VIDEO:

5 minutos sobre historia, razon de ser, uso.
Muestra y explicación de código: 
    Demostrar que sabemos programar en el lenguaje.
    No mostrar cosas triviales (no es necesario código super complejo).
Hablar coloquialmente para hacerlo dinamico y entretenido.


Hacer subsets para las ejecuciones a mostrar en el video y para que lo pueda ejecutar el profe.
Realizar un proceso paralelo para que guarde y lea CSVs (truncados para hacerlos pequeños).
Para hacer eso filtrar por una lista de imdb_ids.

Comparacion OZ con Julia
    Pattern matching

=#

File: prepare_data.py
from numpy import nan
from pandas import DataFrame, concat, isna, Series
from bson.regex import Regex
from datetime import datetime

from mongo import Mongo


log = lambda x: print(f"{datetime.now().replace(microsecond=0)} | {x}")

mongo = Mongo()

db_testing = 'testing'

col_imdb_complete = 'imdb_complete'
col_tmdb_complete = 'tmdb_complete'
col_comments_rating = 'collaborative_db'

col_titlebasics = 'imdbTitleBasics'
col_titleratings = 'imdbTitleRatings'
col_namebasics = 'imdbNameBasics'

col_datatp = 'data_tp'

# Columns that will have the final dataframe
cols = [
    'imdbId',
    'country',
    'releaseDate',
    'votes',
    'rating',
    'reviews',
    'duration',
    'genres',
    'companies',
    'keywords',
    'directors',
    'cast',
]

"""
    imdbId 		: TitleBasics 'imdbId'
    type		: TitleBasics 'Type'
    country 	: _ ''
    releaseDate : _ '
    year		: TitleBasics 'startYear'
    votes 		: imdbTitleRatings 'numVotes'
    rating 		: imdbTitleRatings 'averageRaings'
    reviews 	: collaborative_db 'reviews_analysis.compound'
    duration 	: TitleBasics 'runtimeMinutes'
    genres 		: list TitleBasics 'genres'
    companies 	: list _ ''
    keywords 	: list _ ''
    directors 	: list NameBasics 'primaryName' # Filtrar por 'director' en primaryPosition, abrir por knownForTitles, matchear dejando primaryName
    cast 		: list _ ''
"""

imdbs_list = [
    "tt10133702",
    "tt10333266",
    "tt10350420",
    "tt10399902",
    "tt10402396",
    "tt10469410",
    "tt10712472",
    "tt11244166",
    "tt11257606",
    "tt11281192",
]
query_imdbs = {} # { 'imdbId': { '$in': imdbs_list } }



def get_data_from_mongo(database, collection, query, projection):

    log(f"Searching data in '{database}.{collection}'.")
    df = DataFrame()
    
    total_items = mongo.get_total_documents(
        database=database, collection=collection,
        query=query
    )
    step = 1000000
    log(f"\tThere are {total_items} to get.")

    for start in range(0, total_items+1, step):

        pipeline = [
            {'$skip': start},
            {'$limit': step},
            {'$match': query},
            {'$project': projection},
        ]

        payloads = mongo.get_group_db_item(database=database, collection=collection, query=pipeline)
        df = concat([df, DataFrame(payloads)])
        log(f"\t\t{df.shape[0]} / {total_items}".ljust(100))#, end='\r')

    if 'imdbId' in df.columns:
        df = df.dropna(subset='imdbId')
    df = df.replace(to_replace='\\N', value=None)
    df = df.replace(to_replace='\\\\N', value=None)

    log(f"\t{df.shape[0]} total items got.\n")
    return df



# Imdb TitleBasics data
projection = {
    '_id': 0,
    'imdbId': '$imdbId',
    'type': '$titleType',
    'year': '$startYear',
    'duration': '$runtimeMinutes',
    'genres': '$genres',
}
df_titlebasics = get_data_from_mongo(
    database=db_testing, collection=col_titlebasics, 
    query=query_imdbs, projection=projection
)
df_titlebasics['genres'] = df_titlebasics['genres'].apply(lambda x: x.split(',') if not ( isna(x) or x is None ) else None)
log(f"{df_titlebasics.shape[0]} items from {col_titlebasics} \n{df_titlebasics.head(2)}\n")
df = df_titlebasics.copy() # Columns: imdbId, type, year, duration, genres
log(f"\t -->> {df.shape[0]} ITEMS <<--\n")
del df_titlebasics


# Imdb TitleRatings data
projection = {
    '_id': 0,
    'imdbId': '$imdbId',
    'rating': '$averageRating',
    'votes': '$numVotes',
}
df_titleratings = get_data_from_mongo(
    database=db_testing, collection=col_titleratings, 
    query=query_imdbs, projection=projection
)
log(f"{df_titleratings.shape[0]} items from {col_titleratings} \n{df_titleratings.head(2)}\n")
df = df.merge(df_titleratings, how='left', on='imdbId') # Add columns rating, votes
log(f"\t -->> {df.shape[0]} ITEMS <<--\n")
del df_titleratings


# Letterboxd reviews data 
query = {} # { 'imdb_id': { '$in': imdbs_list } }
projection = {
    '_id': 0,
    'imdbId': '$imdb_id',
    'reviews': '$reviews_analysis.compound',
}
df_collaborativedb = get_data_from_mongo(
    database=db_testing, collection=col_comments_rating, 
    query=query, projection=projection
)
df_collaborativedb = df_collaborativedb.groupby('imdbId').agg({'reviews':'max'}).reset_index()
log(f"{df_collaborativedb.shape[0]} items from {col_comments_rating} \n{df_collaborativedb.head(2)}\n")
df = df.merge(df_collaborativedb, how='left', on='imdbId') # Add columns: reviews
log(f"\t -->> {df.shape[0]} ITEMS <<--\n")
del df_collaborativedb


# # Imdb NameBasics data (directors)
query = {} # { '$or': [ { 'knownForTitles':Regex(f".*{v}.*", "i") } for v in imdbs_list ] }
projection = {
    '_id': 0,
    'profession': '$primaryProfession',
    'name': '$primaryName',
    'titles': '$knownForTitles'
}
df_namebasics = get_data_from_mongo(
    database=db_testing, collection=col_namebasics, 
    query=query, projection=projection
)
df_namebasics = df_namebasics[ df_namebasics['profession'].apply( lambda x: 'director' in x if not ( isna(x) or x is None ) else False ) ]
df_namebasics = df_namebasics[[ 'name', 'titles' ]]
df_namebasics['titles'] = df_namebasics['titles'].apply( lambda x: x.split(',') if not ( isna(x) or x is None ) else None )
df_namebasics = df_namebasics.explode(column='titles').rename(columns={'titles':'imdbId'})
# df_namebasics = df_namebasics[ df_namebasics['imdbId'].isin(imdbs_list) ]
df_directors = df_namebasics.groupby('imdbId').agg({'name':list}).reset_index()
df_directors = df_directors.rename(columns={'name':'directors'})
log(f"{df_directors.shape[0]} items from {col_namebasics} \n{df_directors.head(2)}\n")
df = df.merge(df_directors, how='left', on='imdbId') # Add column: directors
log(f"\t -->> {df.shape[0]} ITEMS <<--\n")
del df_directors, df_namebasics


# Imdb complete data 
query = {} # { 'Id': { '$in': imdbs_list } }
projection = {
    '_id': 0,
    'imdbId': '$Id',
    'cast': '$Cast',
    'country': '$Country',
    'releaseDate': '$ReleaseDate',
    'keywords': '$Keywords',
    'companies': '$Companies',
}
df_imdb_complete = get_data_from_mongo(
    database=db_testing, collection=col_imdb_complete, 
    query=query, projection=projection
)
cond_company = df_imdb_complete['companies'].notna()
df_imdb_complete.loc[cond_company, 'companies'] = df_imdb_complete.loc[cond_company, 'companies'].apply( 
     lambda x: [ i['Name'] for i in x if i.get('Name') ] )
cond_country = df_imdb_complete['country'].notna()
df_imdb_complete.loc[cond_country, 'country'] = df_imdb_complete.loc[cond_country, 'country'].apply( lambda x: x[0] )
log(f"{df_imdb_complete.shape[0]} items from {col_imdb_complete} \n{df_imdb_complete.head(2)}\n")
df = df.merge(df_imdb_complete, how='left', on='imdbId') # Add columns: cast, country, releaseDate, keywords, companies
log(f"\t -->> {df.shape[0]} ITEMS <<--\n")
del df_imdb_complete


# Tmdb complete data
projection = {
    '_id': 0,
    'imdbId': 1,
    'releaseDate': '$ReleaseDate',
    'votes': '$vote_count',
    'rating': '$vote_average',
    'duration': '$Duration',
    'genres': '$Genres',
    'country': '$Country',
    'keywords': '$Keywords',
    'cast': '$Cast',
    'companies': '$Companies',
    'directors': '$Directors'
}
df_tmdb = get_data_from_mongo(
    database=db_testing, collection=col_tmdb_complete, 
    query=query_imdbs, projection=projection
)
df_tmdb = df_tmdb.dropna(subset='imdbId')
df_tmdb = df_tmdb.drop_duplicates(subset='imdbId')
cond_company = df_tmdb['companies'].notna()
df_tmdb.loc[cond_company, 'companies'] = df_tmdb.loc[cond_company, 'companies'].apply( 
     lambda x: [ i['Name'] for i in x if i.get('Name') ] )
cond_country = df_tmdb['country'].notna()
df_tmdb.loc[cond_country, 'country'] = df_tmdb.loc[cond_country, 'country'].apply( lambda x: x[0] )
log(f"{df_tmdb.shape[0]} items from {col_tmdb_complete} \n{df_tmdb.head(2)}\n")


log(f"Setting null values")
df = df.fillna(value=nan).replace(to_replace=nan, value=None)
df = df.replace(to_replace='', value=None)
df = df.replace(to_replace='\\N', value=None)
df = df.replace(to_replace='\\\\N', value=None)
log(f"\t -->> {df.shape[0]} ITEMS <<--\n")


def complete_data(df_data, df_tmdb):

    cols = df_data.columns.tolist()

    log(f"Completing data for {df_data.shape[0]} rows:\n")

    fields = [
		'releaseDate',
		'votes',
		'rating',
		'duration',
        'genres',
        'country',
        'keywords',
        'cast',
        'companies',
        'directors',
    ]

    for index, field in enumerate(fields):

        log(f"\tField '{field}' {index+1}/{len(fields)}")

        condition = ( df_data[field].isna() ) | ( df_data[field]=="" )
        null_data = df_data[ condition ]
        if null_data.empty: continue
        not_null_data = df_data[ ~condition ]
        log(f"\t\t{df_data['imdbId'].unique().size}  -  {null_data.shape[0]} null data ({null_data['imdbId'].unique().size})  |  {not_null_data.shape[0]} not null data ({not_null_data['imdbId'].unique().size}) .")

        df_tmdb_info = df_tmdb[['imdbId', field]]
        df_tmdb_info = df_tmdb_info[ ( df_tmdb_info[field].notna() ) & ( df_tmdb_info[field]!="" ) ]

        null_data.drop(columns=field, inplace=True)
        null_data = null_data.merge(df_tmdb_info, on='imdbId', how='left')
        log(f"\t\tData set : ")
        log(f"\t\t{df_data.shape[0]} - {df_data['imdbId'].unique().size}  |  {null_data.shape[0]} - {null_data['imdbId'].unique().size})  |  {not_null_data.shape[0]} - {not_null_data['imdbId'].unique().size}) .")

        df_data = concat([ null_data[cols] , not_null_data[cols] ])
        log(f"\t\tConcatenated : {df_data.shape[0]} rows.\n")
	
    return df_data


df = complete_data(df_data=df, df_tmdb=df_tmdb)
log(f"\t -->> {df.shape[0]} ITEMS <<--\n")
del df_tmdb


log(f"Setting null values")
df = df.fillna(value=nan).replace(to_replace=nan, value=None)
df = df.replace(to_replace='', value=None)
df = df.replace(to_replace='\\N', value=None)
df = df.replace(to_replace='\\\\N', value=None)
log(f"\t -->> {df.shape[0]} ITEMS <<--\n")


i, j, n = 0, 0, df.shape[0]
log(f"Inserting {n} items to mongo")
items_to_insert = list()
for index, row in df.iterrows():
    i += 1
    j += 1
    doc = row.to_dict()
    for k, v in doc.items():
        if v is nan:
            doc[k] = None
    items_to_insert.append(doc)
    if i >= 10000:
        print(f"\t{j} / {n}".ljust(100), end='\r')
        mongo.insert_many_items(
            database=db_testing, 
            collection=col_datatp, 
            payloads=items_to_insert
        )
        items_to_insert.clear()
        i = 0

mongo.insert_many_items(
    database=db_testing, 
    collection=col_datatp, 
    payloads=items_to_insert
)
items_to_insert.clear()


log(f"Deleting values with no rating")
df = df[ df['rating'].notna() ]
log(f"\t -->> {df.shape[0]} ITEMS <<--\n")
log(f"{df.shape[0]} rows to write in csv.")


def write_csv(df, first):
    csv_filename = 'tp_data.csv'
    mode = 'w' if first else 'a'
    df.to_csv(csv_filename, encoding='utf-8', index=False, header=first, mode=mode)


log(f"Writing data to csv")
n, step = df.shape[0], 50000
for i in range(0, n, step):
    a = df.iloc[i:i+step]
    print(f"\t{i+step} / {n}  |  {a.shape[0]} rows".ljust(100), end='\r')
    write_csv( df=a , first=i==0)


mongo.__del__()


log(f"FINISH")
